### ISSUE TITLE
diarization with Streaming?

### ISSUE THREAD

**Issue by wavepurrticle-create** [altunenes/parakeet-rs#41]
Hey, is there a way to have this work in the ultra low latency ? this would help with real time streaming.
Config | Chunk Size | Latency | RTF
-- | -- | -- | --
Ultra Low | 3 frames | 0.32s | 0.180
Low | 6 frames | 1.04s | 0.093
High | 124 frames | 10.0s | 0.005
Very High | 340 frames | 30.4s | 0.002
@altunenes 
as described here?  
https://huggingface.co/nvidia/diar_streaming_sortformer_4spk-v2

**Comment by altunenes** [comment]
our current config (124 frames, ~10s latency) is actually NVIDIA's recommended cfg. 🙂
Going to ultra-low (3 frames) would hurt acc since the model was trained around 188 frames, and the compute cost jumps 36x for something that's already pretty fast (RTF 0.005). The 10s is just buffer latency anyway processing is near real-time. 

To do what you want, I need to get different ONNX exports for each config in this table. We could export 4 different onnx models for that. but supporting multiple configs also means maintaining separate ONNX exports 🙄  and I'm not sure how much that is worth. Because, as I said, the values we are currently using are already recommended values that give fast and acc  🙂

**Comment by MathieuDuponchelle** [comment]
> I love gst and use it in many cases so happy to help.

Yes I noticed you mentioned it elsewhere :) I'll make sure to link the merge request here once I get to it so that you can review the relevant bits.

> Just to make sure we're on the same page: making the Rust constants configurable is only half of it. You'd also need a separate ONNX export for each chunk config (the chunk size is baked into the ONNX graph at export time). So it's really a two-part change: different ONNX models + matching Rust config.

Yes, that's how I understood it, by the way the ideal situation would be the ability to look up the chunk config from the ONNX model, that would entirely eliminate user error, do you think it is possible?

>  Do you need my ONNX export script for this? I'll need to set it up. please let me know thank you

Yes :)

**Comment by altunenes** [comment]
> Yes, that's how I understood it, by the way the ideal situation would be the ability to look up the chunk config from the ONNX model, that would entirely eliminate user error, do you think it is possible?

Not from the graph directly. since my export uses dynamic axes, the actual values aren't stored in the shapes. But ONNX supports custom model metadata (key value strings). We could embed chunk_len/fifo_len/cache_len at export time, then read them on the Rust side. That would be a **probably** clean way to make the config self describing. Good idea actually 

Here is my ONNX export. I ran it on Google Colab env. It should be working right now also no any major changes relelated with torch etc. If it doesn't work, please let me know (sometimes minor errors can occur due to internal updates).

```python
import torch
import types
from nemo.collections.asr.models import SortformerEncLabelModel

print(f"PyTorch version: {torch.__version__}")
# or v2.nemo, its same arch. 
MODEL_PATH = "diar_streaming_sortformer_4spk-v2.1.nemo"
ONNX_OUTPUT_PATH = "diar_streaming_sortformer_4spk-v2.1.onnx"

print(f”loadinf model form: {MODEL_PATH}")
model = SortformerEncLabelModel.restore_from(restore_path=MODEL_PATH, map_location='cpu')
model.eval()

CHUNK_LEN = 124
FIFO_LEN = 124
CACHE_LEN = 188
FEAT_DIM = 128
EMB_DIM = 512

model.sortformer_modules.chunk_len = CHUNK_LEN
model.sortformer_modules.chunk_right_context = 1
model.sortformer_modules.fifo_len = FIFO_LEN
model.sortformer_modules.spkcache_update_period = CHUNK_LEN
model.sortformer_modules.spkcache_len = CACHE_LEN
model.sortformer_modules._check_streaming_parameters()

def onnx_forward(self, chunk, chunk_lengths, spkcache, spkcache_lengths, fifo, fifo_lengths):
    chunk_pre_encode_embs, chunk_pre_encode_lengths = self.encoder.pre_encode(x=chunk, lengths=chunk_lengths)
    chunk_pre_encode_lengths = chunk_pre_encode_lengths.to(torch.int64)

    concat_embs = torch.cat([spkcache, fifo, chunk_pre_encode_embs], dim=1)
    concat_lens = spkcache_lengths + fifo_lengths + chunk_pre_encode_lengths

    spkcache_fifo_chunk_fc_encoder_embs, spkcache_fifo_chunk_fc_encoder_lengths = self.frontend_encoder(
        processed_signal=concat_embs,
        processed_signal_length=concat_lens,
        bypass_pre_encode=True,
    )

    spkcache_fifo_chunk_preds = self.forward_infer(
        spkcache_fifo_chunk_fc_encoder_embs, spkcache_fifo_chunk_fc_encoder_lengths
    )

    return spkcache_fifo_chunk_preds, chunk_pre_encode_embs, chunk_pre_encode_lengths

model.forward = types.MethodType(onnx_forward, model)

batch_size = 1
subsampling = 8
chunk_frames_in = CHUNK_LEN * subsampling

chunk = torch.randn(batch_size, chunk_frames_in, FEAT_DIM)
chunk_lengths = torch.tensor([chunk_frames_in], dtype=torch.long)
spkcache = torch.zeros(batch_size, CACHE_LEN, EMB_DIM)
spkcache_lengths = torch.tensor([0], dtype=torch.long)
fifo = torch.zeros(batch_size, FIFO_LEN, EMB_DIM)
fifo_lengths = torch.tensor([0], dtype=torch.long)

input_example = (chunk, chunk_lengths, spkcache, spkcache_lengths, fifo, fifo_lengths)

print(f"  chunk    {chunk.shape}")
print(f"  spkcahe: {spkcache.shape}")
print(f"  fif:     {fifo.shape}")

torch.onnx.export(
    model,
    input_example,
    ONNX_OUTPUT_PATH,
    input_names=["chunk", "chunk_lengths", "spkcache", "spkcache_lengths", "fifo", "fifo_lengths"],
    output_names=["spkcache_fifo_chunk_preds", "chunk_pre_encode_embs", "chunk_pre_encode_lengths"],
    dynamic_axes={
        "chunk": {0: "batch", 1: "time_chunk"},
        "spkcache": {0: "batch", 1: "time_cache"},
        "fifo": {0: "batch", 1: "time_fifo"},
        "spkcache_fifo_chunk_preds": {0: "batch", 1: "time_out"},
        "chunk_pre_encode_embs": {0: "batch", 1: "time_pre_encode"}
    },
    opset_version=17,
    dynamo=False,
    verbose=False,
)

print(f"xported to: {ONNX_OUTPUT_PATH}")


import onnx
model_onnx = onnx.load(ONNX_OUTPUT_PATH)
print("\n verify input shapes:")
for inp in model_onnx.graph.input:
    dims = [d.dim_param if d.dim_param else d.dim_value for d in inp.type.tensor_type.shape.dim]
    print(f"  {inp.name}: {dims}")

fifo_input = [inp for inp in model_onnx.graph.input if inp.name == "fifo"][0]
fifo_dim1 = fifo_input.type.tensor_type.shape.dim[1]
if fifo_dim1.dim_param:
    print(f"\n FIFO is dynamic: '{fifo_dim1.dim_param}'")
else:
    print(f"\n FIFO is fixed: {fifo_dim1.dim_value}")

```

**Comment by MathieuDuponchelle** [comment]
Sorry I was about to answer then power cut here :)

> btw this should work for what you think. it doesn't touch the graph at all so more cfg options in rust side.

On rust side `Sortformer` would attempt to read those directly right?

Those test results are really encouraging, if I understand correctly half-size will result in a 5-second latency? The deviations between configurations seem acceptable, as long as the results are still decently aligned with the actual transcripts we should be able to get something nice cooking :)

For what it's worth my ultimate goal is to implement everything needed inference-wise to reproduce the results I presented at the last GStreamer conference (gstconf.ubicast.tv/videos/costly-speech-an-introduction/), but using only local models with no service fees. For that demonstration I used speechmatics for transcription and diarization, awstranslate for translation and elevenlabs for voice cloning / speech synthesis.

At this stage there is now a whisper transcriber upstream that gives good results with fairly low latency, diarization is as far as I can tell the least well-solved piece of the puzzle, but in my testing sortformer will be more than adequate for low-speaker-count conversation use cases.

My extended use case would be supporting council-meeting-type audio, where any number of speakers may come and go, but I'll already be thrilled once I get a completely local pipeline for the simpler use case :)

**Comment by MathieuDuponchelle** [comment]
> The main limitation right now is the 4-speaker cap in v2/v2.1

Yes, that's also my concern, but I'd rather something that works well, fast, and with a clear API for streaming, I think I also read that nvidia were working on an 8-speaker version. I also noticed what you suggested about using dedicated embedding models in combination, but I haven't done any research into that, to be honest my main concern is ending up with simple-to-maintain elements upstream :)

> The values are read from ONNX metadata at load time, but the fields are pub so you can override after construction. Does this style work for your gst use case, or do you think there's a better API shape?

With parakeet taking care of reading the values, I don't think I will expose configuration options, instead what would be best is a way to query the sortformer for its latency after construction with a particular ONNX export. If that's possible, I will simply expose a property to let the user specify the path to an export, query the latency from sortformer and advertise, and consider my work done :)

sağ ol :)

**Comment by MathieuDuponchelle** [comment]
> Worth noting that NVIDIA's own team confirmed you can experiment with different chunk sizes without changing other parameters, though accuracy may decrease when deviating too far from the training setup (chunk size of 188 frames), please see: https://huggingface.co/nvidia/diar_streaming_sortformer_4spk-v2/discussions/6.

Ah I see, I was wondering what the source was for your comment @ https://github.com/altunenes/parakeet-rs/issues/41#issuecomment-3623797878, thanks :)

I will also experiment with custom exports and report results, hopefully next week!
